In [15]:
import pandas as pd

# read xlsx file

df_baby_out_of_hospital = pd.read_excel('data.xlsx', sheet_name='UOHB Neonates Characteristics')
df_baby_in_hospital = pd.read_excel('data.xlsx', sheet_name='IHB Neonates Characteristics')

column_name_mapping = {
    'OutofHospital': 'OutofHospital',
    'level_1': 'level_1',
    'Birth weight (g)': 'Birth weight',
    'Gestational Age at Birth': 'Gestational Age at Birth',
    'APGAR score (5 min.)': 'APGAR score',
    'Neonatal Survival (up to 28 days) (0 = deceased, 1 = alive, N/A for lost)': 'Neonatal Survival',
    'Perinatal Survival (up to 154 days) (0 = deceased, 1 = alive, N/A for lost)': 'Perinatal Survival',
    'Polycythemia (0 = no, 1 = yes)': 'Polycythemia',
    'Hypoglycemia (0 = no, 1 = yes)': 'Hypoglycemia',
    'Convulsions (0 = no, 1 = yes)': 'Convulsions',
    'Hypothermic (0 = no, 1 = yes)': 'Hypothermic',
    'Bradycardia (0 = no, 1 = yes)': 'Bradycardia',
    'Respiratory distress (0 = no, 1 = yes)': 'Respiratory distress',
    'PPHN (0 = no, 1 = yes)': 'PPHN',
    'Metabolic/respiratory acidosis (0 = no, 1 = yes)': 'Metabolic/respiratory acidosis',
    'Other (list what)': 'Other',
    '18mo Developmental Delay (0 = NOT LISTED/LOST, 1 = none, 2 = yes)': '18mo Developmental Delay',
    '18mo pathologies (0 = NOT LISTED/LOST, 1 = none, 2 = yes)': '18mo pathologies',
    'Neonatal Survival (up to 28 days) (0 = deceased, 1 = alive)': 'Neonatal Survival',
    'Perinatal Survival (up to 154 days) (0 = deceased, 1 = alive)': 'Perinatal Survival',
    'Metabolic/Respiratory Acidosis (0 = no, 1 = yes)': 'Metabolic/Respiratory Acidosis'
}


# rename column to "Pregnancy Complications"
df_baby_out_of_hospital.rename(columns=column_name_mapping, inplace=True)
df_baby_in_hospital.rename(columns=column_name_mapping, inplace=True)

df = pd.concat([df_baby_out_of_hospital, df_baby_in_hospital], keys=['Out of Hospital', 'In Hospital'])
df.reset_index(inplace=True)
df.rename(columns={'level_0': 'OutofHospital'}, inplace=True)
df.rename(columns={'level_1': 'InHospital'}, inplace=True)
df['OutofHospital'] = df['OutofHospital'].apply(lambda x: 1 if x == 'Out of Hospital' else 0)

# Remove InHopital column
df.drop(columns=['InHospital'], inplace=True)


In [16]:
df

OutofHospital  Birth weight Gestational Age at Birth  APGAR score  \
0                1          2290                36.571429          9.0   
1                1          2400                37.571429          9.0   
2                1          3100                38.428571          NaN   
3                1          2610                      NaN          7.0   
4                1          3103                39.428571          9.0   
..             ...           ...                      ...          ...   
133              0          2665                       39          9.0   
134              0          3020                       39          9.0   
135              0          3580                39.142857          9.0   
136              0          4595                40.857143          9.0   
137              0          2985                39.285714          9.0   

     Neonatal Survival  Perinatal Survival  Polycythemia  Hypoglycemia  \
0                  1.0                 1.0             0             0   
1                  NaN                 NaN             0             0   
2                  1.0                 1.0             0             0   
3                  1.0                 1.0             0             0   
4                  NaN                 NaN             0             0   
..                 ...                 ...           ...           ...   
133                1.0                 1.0             0             0   
134                1.0                 NaN             0             0   
135                1.0                 1.0             0             0   
136                NaN                 NaN             0             0   
137                1.0                 1.0             0             0   

     Convulsions  Hypothermic  Bradycardia  Respiratory distress  PPHN  \
0              0            0            0                     0     0   
1              0            0            0                     0     0   
2              0            0            0                     0     0   
3              0            0            0                     0     0   
4              0            0            1                     0     0   
..           ...          ...          ...                   ...   ...   
133            0            0            0                     0     0   
134            0            0            0                     0     0   
135            0            0            0                     0     0   
136            0            0            0                     0     0   
137            0            0            0                     1     0   

     Metabolic/respiratory acidosis                               Other  \
0                               0.0                                 SGA   
1                               0.0                                 SGA   
2                               0.0                  Perioneal cyanosis   
3                               0.0                                 SGA   
4                               0.0                  Hyperbilirubinemia   
..                              ...                                 ...   
133                             NaN                  ABO, SGA, meconium   
134                             NaN    microcephaly,  Caput succedaneum   
135                             NaN                              murmur   
136                             NaN                               LGA     
137                             NaN  ABO, murmur, RSV, thrombocytopenia   

     18mo Developmental Delay  18mo pathologies  \
0                         0.0               0.0   
1                         0.0               0.0   
2                         0.0               0.0   
3                         1.0               1.0   
4                         0.0               0.0   
..                        ...               ...   
133                       0.0               0.0   
134                       0.0

In [20]:
mapping_dictionaries = {
    'OutofHospital': {
        0: 'no',
        1: 'yes'
    },
    'Neonatal Survival': {
        0: 'deceased',
        1: 'alive',
        'N/A': 'lost'
    },
    'Perinatal Survival': {
        0: 'deceased',
        1: 'alive',
        'N/A': 'lost'
    },
    'Polycythemia': {
        0: 'no',
        1: 'yes'
    },
    'Hypoglycemia': {
        0: 'no',
        1: 'yes'
    },
    'Convulsions': {
        0: 'no',
        1: 'yes'
    },
    'Hypothermic': {
        0: 'no',
        1: 'yes'
    },
    'Bradycardia': {
        0: 'no',
        1: 'yes'
    },
    'Respiratory distress': {
        0: 'no',
        1: 'yes'
    },
    'PPHN': {
        0: 'no',
        1: 'yes'
    },
    'Metabolic/respiratory acidosis': {
        0: 'no',
        1: 'yes'
    },
    '18mo Developmental Delay': {
        0: 'not listed/lost',
        1: 'none',
        2: 'yes'
    },
    '18mo pathologies': {
        0: 'not listed/lost',
        1: 'none',
        2: 'yes'
    },
    'Neonatal Survival': {
        0: 'deceased',
        1: 'alive'
    },
    'Perinatal Survival': {
        0: 'deceased',
        1: 'alive'
    },
    'Metabolic/Respiratory Acidosis': {
        0: 'no',
        1: 'yes'
    },
}


# Replacing numerical values with categories for each column in the DataFrame
for column, mapping in mapping_dictionaries.items():
    df[column] = df[column].replace(mapping)

## Chi-Square Test of Independence

This test is used to determine if there is a significant association between two categorical variables. In our case, we can create a Rx2 contingency table where one axis represents the location of birth (in hospital or out of hospital) and the other axis represents whether there was a birth complication (yes or no). Then, we can apply the Chi-Square Test to determine if the occurrence of birth complications is independent of the location of birth.

In [21]:
df.columns

Index(['OutofHospital', 'Birth weight', 'Gestational Age at Birth',
       'APGAR score', 'Neonatal Survival', 'Perinatal Survival',
       'Polycythemia', 'Hypoglycemia', 'Convulsions', 'Hypothermic',
       'Bradycardia', 'Respiratory distress', 'PPHN',
       'Metabolic/respiratory acidosis', 'Other', '18mo Developmental Delay',
       '18mo pathologies', 'Metabolic/Respiratory Acidosis'],
      dtype='object')

In [22]:
df

OutofHospital  Birth weight Gestational Age at Birth  APGAR score  \
0             yes          2290                36.571429          9.0   
1             yes          2400                37.571429          9.0   
2             yes          3100                38.428571          NaN   
3             yes          2610                      NaN          7.0   
4             yes          3103                39.428571          9.0   
..            ...           ...                      ...          ...   
133            no          2665                       39          9.0   
134            no          3020                       39          9.0   
135            no          3580                39.142857          9.0   
136            no          4595                40.857143          9.0   
137            no          2985                39.285714          9.0   

    Neonatal Survival Perinatal Survival Polycythemia Hypoglycemia  \
0               alive              alive           no           no   
1                 NaN                NaN           no           no   
2               alive              alive           no           no   
3               alive              alive           no           no   
4                 NaN                NaN           no           no   
..                ...                ...          ...          ...   
133             alive              alive           no           no   
134             alive                NaN           no           no   
135             alive              alive           no           no   
136               NaN                NaN           no           no   
137             alive              alive           no           no   

    Convulsions Hypothermic Bradycardia Respiratory distress PPHN  \
0            no          no          no                   no   no   
1            no          no          no                   no   no   
2            no          no          no                   no   no   
3            no          no          no                   no   no   
4            no          no         yes                   no   no   
..          ...         ...         ...                  ...  ...   
133          no          no          no                   no   no   
134          no          no          no                   no   no   
135          no          no          no                   no   no   
136          no          no          no                   no   no   
137          no          no          no                  yes   no   

    Metabolic/respiratory acidosis                               Other  \
0                               no                                 SGA   
1                               no                                 SGA   
2                               no                  Perioneal cyanosis   
3                               no                                 SGA   
4                               no                  Hyperbilirubinemia   
..                             ...                                 ...   
133                            NaN                  ABO, SGA, meconium   
134                            NaN    microcephaly,  Caput succedaneum   
135                            NaN                              murmur   
136                            NaN                               LGA     
137                            NaN  ABO, murmur, RSV, thrombocytopenia   

    18mo Developmental Delay 18mo pathologies Metabolic/Respiratory Acidosis  
0            not listed/lost  not listed/lost                            NaN  
1            not listed/lost  not listed/lost                            NaN  
2            not listed/lost  not listed/lost                            NaN  
3                       none             none                            NaN  
4            not listed/lost  not listed/lost                            NaN  
..                       ...              ...                            ...  
133          no

In [25]:
import pandas as pd
from scipy.stats import chi2_contingency, fisher_exact

# List of variables to check for association with 'OutofHospital'
variables = ["Neonatal Survival", "Perinatal Survival", "Polycythemia", "Hypoglycemia", "Convulsions", "Hypothermic", "Bradycardia", "Respiratory distress", "PPHN"]

# Loop through each variable and perform the Chi-Square Test
results = {}
for variable in variables:
    # Create a contingency table
    contingency_table = pd.crosstab(df['OutofHospital'], df[variable])
    
    # Perform the Chi-Square Test of Independence
    chi2, p, dof, expected = chi2_contingency(contingency_table)
    
    # Store the results
    results[variable] = {'Chi-Square Value': chi2, 'P-Value': p, 'Degrees of Freedom': dof, 'Expected Frequencies Table': expected}

# Output the results in a beautiful way
print("{:<25} {:<20} {:<20} {:<20}".format("Variable", "Chi-Square Value", "P-Value", "Degrees of Freedom"))
for variable, result in results.items():
    print("{:<25} {:<20} {:<20} {:<20}".format(variable, result['Chi-Square Value'], result['P-Value'], result['Degrees of Freedom']))


Variable                  Chi-Square Value     P-Value              Degrees of Freedom  
Neonatal Survival         0.034177715101852385 0.8533293086053286   1                   
Perinatal Survival        0.06751054852320676  0.7949965125940222   1                   
Polycythemia              3.698000683526999    0.05447770916113348  1                   
Hypoglycemia              24.555285586124405   7.220799970750491e-07 1                   
Convulsions               0.0                  1.0                  1                   
Hypothermic               23.05227272727273    1.5765603373793762e-06 1                   
Bradycardia               4.831597222222223    0.027942732316554454 1                   
Respiratory distress      0.8262186004784692   0.3633682441225963   1                   
PPHN                      1.8562758264462815   0.1730552393175499   1                   


### Interpretation:

* Chi-Square Value: A larger value indicates a greater difference between the observed counts and what would be expected if the variables were independent.

* P-Value: Tells you whether or not the association is statistically significant. Typically, if the p-value is less than 0.05, we might conclude that there is a significant association between the variables.

* Degrees of Freedom: This is equal to (number of rows - 1) x (number of columns - 1) in the contingency table. It helps in determining the critical value for the chi-square distribution.

For interpretation, we would usually focus on the P-Value. If it's below a significance level (commonly 0.05), it suggests that the observed distribution is significantly different from what you'd expect if the variables were independent, indicating an association.

For the variable `Hypoglycemia`, `Hypothermic`, `Polycythemia` and `Bradycardia`, the very low p-value indicates that there is a statistically significant association between race and whether the birth occurred in or out of the hospital. This suggests that race and the language might play a role in the choice or circumstances of the birth location.

For the variable `Neonatal Survival`, the p-value is much greater than 0.05, suggesting that there is no statistically significant association between pregnancy complications and birth location. Pregnancy complications do not appear to be a determining factor in whether a birth occurs in or out of the hospital.


In [51]:
import pandas as pd
import plotly.express as px

# Assuming your dataframe is named df, and has columns 'OutofHospital', 'PrimaryLanguage' and 'Ethnicity'

# Plot for PrimaryLanguage
fig1 = px.histogram(df, x="Hypothermic", color="OutofHospital",
                    barmode="group", nbins=len(df["Hypothermic"].unique()),
                    labels={"PrimaryLanguage": "Primary Language", "OutofHospital": "Out of Hospital"},
                    title="Number of Births In and Out of Hospital by Primary Language")

fig1.show()


# Plot for Ethnicity
fig2 = px.histogram(df, x="Hypoglycemia", color="OutofHospital",
                    barmode="group", nbins=len(df["Hypoglycemia"].unique()),
                    labels={"Ethnicity": "Ethnicity", "OutofHospital": "Out of Hospital"},
                    title="Number of Births In and Out of Hospital by Ethnicity",
)

fig2.show()

# export plotly figures to html
import plotly.io as pio
pio.write_html(fig1, file='fig/baby/hypothermic.html', auto_open=False)
pio.write_html(fig2, file='fig/baby/hypoglycemia.html', auto_open=False)


In [50]:
import pandas as pd
import plotly.graph_objects as go
import plotly.io as pio

# Assuming your dataframe is named df, and has columns 'OutofHospital', 'Race', 'Age', 'PrimaryLanguage', 'Ethnicity', 'Education', 'Employment', 'PregnancyComplications', 'Parity'

# List of variables to plot
variables = ["Birth weight", "Polycythemia", "Hypoglycemia", "Convulsions", "Hypothermic", "Bradycardia", "Respiratory distress", "PPHN"]

# Function to create bar chart for a specific variable
def create_trace(variable):
    count_data = df.groupby([variable, 'OutofHospital']).size().reset_index(name='count')
    return {
        'x': count_data[count_data['OutofHospital'] == "yes"][variable],
        'y': count_data[count_data['OutofHospital'] == "yes"]['count'],
        'name': 'Out of Hospital',
        'type': 'bar'
    }, {
        'x': count_data[count_data['OutofHospital'] == "no"][variable],
        'y': count_data[count_data['OutofHospital'] == "no"]['count'],
        'name': 'In Hospital',
        'type': 'bar'
    }

# Create figure
fig = go.Figure()

# Add traces for each variable
for var in variables:
    trace1, trace2 = create_trace(var)
    fig.add_trace(trace1)
    fig.add_trace(trace2)

# Create a dropdown menu
buttons = []
for i, var in enumerate(variables):
    visibility = [False] * len(variables) * 2
    visibility[i*2] = True
    visibility[i*2 + 1] = True
    buttons.append(dict(label=var,
                        method="update",
                        args=[{"visible": visibility},
                              {"title": f"Number of Births In and Out of Hospital by {var}",
                               "xaxis": {"title": var},
                               "yaxis": {"title": "Number of Births"}}]))

# Update layout with dropdown
fig.update_layout(
    updatemenus=[
        go.layout.Updatemenu(
            active=3,  # Set default view to 'Ethnicity'
            buttons=buttons,
            direction="down",
            pad={"r": 10, "t": 10},
            showactive=True,
            x=0.7,
            xanchor="left",
            y=1.3,
            yanchor="top"
        ),
    ],
    barmode="group"
)

# Set initial title
fig.update_layout(title=f"Number of Births In and Out of Hospital by Ethnicity")

# Set initial visibility
visibility = [False] * len(variables) * 2
visibility[3*2] = True
visibility[3*2 + 1] = True
fig.update_traces(visible=False)
fig.data[3*2].visible = True
fig.data[3*2 + 1].visible = True

# Show figure
fig.show()

# Write to HTML
pio.write_html(fig, file='fig/baby/bigfig.html', auto_open=False)

In [48]:
import plotly.express as px

# assuming `df` is your DataFrame

fig = px.box(df, x='OutofHospital', y='Birth weight', 
             labels={"OutofHospital": "Out of Hospital", "Birth weight": "Birth Weight (g)"},
             title="Birth Weight Distribution for Babies Born Inside vs. Outside Hospitals")

fig.show()

# export plotly figures to html
import plotly.io as pio
pio.write_html(fig, file='fig/baby/boxplot1.html', auto_open=False)


In [49]:
fig = px.violin(df, x='OutofHospital', y='Birth weight', 
                labels={"OutofHospital": "Out of Hospital", "Birth weight": "Birth Weight (g)"},
                title="Birth Weight Distribution for Babies Born Inside vs. Outside Hospitals",
                box=True, points="all")

fig.show()

# export plotly figures to html
import plotly.io as pio
pio.write_html(fig, file='fig/baby/violin1.html', auto_open=False)

## Logistic Regresion

Logistic regression is a statistical model that uses a logistic function to model a binary dependent variable. In the context of predicting whether a baby is born in or out of a hospital, it provides a method to estimate the probability of an event (birth location) given a set of related variables, such as race, age, primary language, ethnicity, and education of the parent. Features are assigned coefficients, which reflect the strength and direction of their relationship with the outcome. A positive coefficient suggests that as the feature value increases, the likelihood of the event (in this case, a baby being born out of hospital) also increases, while a negative coefficient suggests the opposite. Thus, logistic regression allows us to identify and quantify the influence of various factors on the outcome, providing insights into relevant and correlated features.

In [30]:
df.columns

Index(['OutofHospital', 'Birth weight', 'Gestational Age at Birth',
       'APGAR score', 'Neonatal Survival', 'Perinatal Survival',
       'Polycythemia', 'Hypoglycemia', 'Convulsions', 'Hypothermic',
       'Bradycardia', 'Respiratory distress', 'PPHN',
       'Metabolic/respiratory acidosis', 'Other', '18mo Developmental Delay',
       '18mo pathologies', 'Metabolic/Respiratory Acidosis'],
      dtype='object')

In [42]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import OneHotEncoder, StandardScaler

# Load data
# Assuming that the dataframe is already loaded in variable df

# Preprocessing
# Select variables of interest
variables_of_interest = ["Birth weight", "Neonatal Survival", "Perinatal Survival", "Polycythemia", "Hypoglycemia", "Convulsions", "Hypothermic", "Bradycardia", "Respiratory distress", "PPHN"]
X = df[variables_of_interest]

# One hot encoding for categorical variables
categorical_vars = ["Neonatal Survival", "Perinatal Survival", "Polycythemia", "Hypoglycemia", "Convulsions", "Hypothermic", "Bradycardia", "Respiratory distress", "PPHN"]
one_hot_encoder = OneHotEncoder(drop='first')
one_hot_encoded = one_hot_encoder.fit_transform(X[categorical_vars])
encoded_features = one_hot_encoder.get_feature_names_out(categorical_vars)
encoded_df = pd.DataFrame(one_hot_encoded.toarray(), columns=encoded_features)

# Combine the one-hot-encoded variables with the numeric variables
X = pd.concat([X.drop(categorical_vars, axis=1), encoded_df], axis=1)

# Standardize numerical features
scaler = StandardScaler()
X['Birth weight'] = scaler.fit_transform(X[['Birth weight']])

# Target variable
y = df['OutofHospital']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Logistic regression|
log_reg = LogisticRegression(max_iter=1000)
log_reg.fit(X_train, y_train)

# Evaluate feature importance
feature_importance = pd.DataFrame({"Feature": X.columns, "Coefficient": log_reg.coef_[0]})
feature_importance = feature_importance.sort_values(by="Coefficient", key=abs, ascending=False).reset_index(drop=True)

print("Feature Importances:")
print(feature_importance)

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Predictions on the test set
y_pred = log_reg.predict(X_test)

# Turn y_pred and y_test into binary
y_pred = np.where(y_pred == 'yes', 1, 0)
y_test = np.where(y_test == 'yes', 1, 0)

# Calculate metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print metrics
print(f'Accuracy: {accuracy:.2f}')
print(f'Precision: {precision:.2f}')
print(f'Recall: {recall:.2f}')
print(f'F1 Score: {f1:.2f}')


Feature Importances:
                        Feature  Coefficient
0               Hypothermic_yes     1.804393
1              Hypoglycemia_yes     1.671158
2         Neonatal Survival_nan     0.686115
3                      PPHN_yes     0.651477
4                  Birth weight    -0.540175
5               Convulsions_yes    -0.360674
6        Perinatal Survival_nan     0.354218
7               Bradycardia_yes     0.247487
8              Polycythemia_yes     0.204884
9      Respiratory distress_yes    -0.048227
10   Neonatal Survival_deceased     0.000000
11  Perinatal Survival_deceased     0.000000
Accuracy: 0.79
Precision: 1.00
Recall: 0.57
F1 Score: 0.73


This study employs a logistic regression model to identify factors that can predict whether a baby will be born in or out of hospital. The results are represented by the coefficients assigned to each feature or variable in the model, which represent the log odds of a unit change in the respective feature, while all other features are held constant. A positive coefficient indicates that the feature is associated with an increased likelihood of the outcome (i.e., a baby being born in hospital), while a negative coefficient suggests a reduced likelihood.

The following factors significantly increase the odds of a baby being born in hospital:

1. Hypothermia (coefficient 1.804): This is the most influential factor, implying that the odds of a baby being born in hospital increases substantially if the baby experiences hypothermia.

2. Hypoglycemia (coefficient 1.671): Similarly, this is a significant predictor, indicating that hypoglycemia also considerably increases the likelihood of in-hospital birth.

3. Unavailable data on neonatal survival (coefficient 0.686) and Persistent Pulmonary Hypertension of the Newborn (PPHN) (coefficient 0.651) also have a positive influence, albeit less strong.

On the contrary, some factors decrease the odds of a baby being born in hospital:

1. Birth weight (coefficient -0.540): This result suggests that an increase in birth weight is associated with a decrease in the odds of an in-hospital birth.

2. Convulsions (coefficient -0.360): The presence of convulsions also reduces the likelihood of in-hospital birth.

Of note, the model accuracy, which represents the proportion of correct predictions made out of all predictions, is 79%, suggesting a reasonable fit to the data. Precision is at 1.00, meaning that all positive predictions made by the model (i.e., all predictions of in-hospital births) were correct. However, the model's recall (also known as sensitivity) is 0.57, indicating that it identified just over half of the actual in-hospital births correctly. Therefore, while the model is highly reliable when it predicts an in-hospital birth, it is relatively less successful in capturing all such cases. This is further reflected in the F1 score, which is a harmonic mean of precision and recall, and is 0.73.

In conclusion, hypothermia and hypoglycemia are significant predictors of a baby being born in hospital. An increase in birth weight and the presence of convulsions reduce the odds of an in-hospital birth. The model is fairly accurate and highly precise, but its sensitivity is moderate. Further research is needed to improve the model's ability to correctly identify all cases of in-hospital births.

In [52]:
import plotly.express as px

# Feature Importance
feature_importance = pd.DataFrame({"Feature": X.columns, "Coefficient": log_reg.coef_[0]})
feature_importance = feature_importance.sort_values(by="Coefficient")

fig = px.bar(feature_importance, x='Coefficient', y='Feature', orientation='h',
             title='Feature Importances',
             labels={'Coefficient':'Coefficient', 'Feature':'Feature'},
             color='Coefficient',
             color_continuous_scale=px.colors.diverging.RdBu,
             height=800)

fig.show()

# export to html
pio.write_html(fig, file='fig/baby/feature_importance.html', auto_open=False)

